### ISSUE TITLE
mastR

### ISSUE THREAD

**Issue by Gene233** [Bioconductor/Contributions#2975]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/DavisLaboratory/mastR

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
I got 
```
This typically indicates Rd problems.
LaTeX errors found:
! Package inputenc Error: Unicode character γ (U+03B3)
(inputenc)                not set up for use with LaTeX.

```
suggesting you have unicode in text instead of latex equivalent (e.g., \gamma).

**Comment by Gene233** [comment]
> I got
> 
> ```
> This typically indicates Rd problems.
> LaTeX errors found:
> ! Package inputenc Error: Unicode character γ (U+03B3)
> (inputenc)                not set up for use with LaTeX.
> ```
> 
> suggesting you have unicode in text instead of latex equivalent (e.g., \gamma).

Hi @vjcitn , I've replaced the unicode with ASCII. Thank you for your reminding! Hope it works now.

Cheers,
Jinjin

**Comment by Gene233** [comment]
Hi @HelenaLC , I subscribed to the bioc-devel mailing list and tried to follow the guidelines to push and launch a new build. But it failed when I used `git fetch --all`. It asks me for my passphrase and I tried with my remote token passphrase but it failed. I also tried entering the passphrase with nothing, but it still failed. I got the following error message:
```
Fetching origin
Fetching upstream
Enter passphrase for key '/home/users/allstaff/chen.j/.ssh/id_rsa': 
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
error: Could not fetch upstream
```
Could you please tell me what could be wrong with this if you have experienced this before?

**Comment by HelenaLC** [comment]
Hm. Have you followed the instructions [here](https://contributions.bioconductor.org/git-version-control.html#new-package-workflow) re setting up remotes and SSH credentials? Unrelated: not sure why you want to fetch, as there are no changes upstream at this point (but maybe I'm missing something, not sure...)?

**Comment by Gene233** [comment]
> Hm. Have you followed the instructions [here](https://contributions.bioconductor.org/git-version-control.html#new-package-workflow) re setting up remotes and SSH credentials? Unrelated: not sure why you want to fetch, as there are no changes upstream at this point (but maybe I'm missing something, not sure...)?

Hi @HelenaLC , thank you for your reply!

Yes, I followed this instructions and added upstream remote branch to my git.
```
$git remote -v
origin  https://github.com/DavisLaboratory/mastR.git (fetch)
origin  https://github.com/DavisLaboratory/mastR.git (push)
upstream        git@git.bioconductor.org:packages/mastR.git (fetch)
upstream        git@git.bioconductor.org:packages/mastR.git (push)
```
I noticed there're some notes so I wanna modify a bit and push it. As the guideline suggests, I need to fetch all before I make commit and push. Not sure if it's necessary so far? Do I only need to commit and push to my own git repo now and push to bioc repo only after release?

**Comment by lshep** [comment]
You could follow some steps here: http://contributions.bioconductor.org/git-version-control.html#faq  that provide help for different scenarios.  I would suggest creating a new ssh keys that doesn't require a passphrase and adding it to BiocCredentials.  You will need to push to git.bioconductor.org from here on out. To trigger a new build it must contain a valid version bump

**Comment by Gene233** [comment]
> You could follow some steps here: http://contributions.bioconductor.org/git-version-control.html#faq that provide help for different scenarios. I would suggest creating a new ssh keys that doesn't require a passphrase and adding it to BiocCredentials. You will need to push to git.bioconductor.org from here on out. To trigger a new build it must contain a valid version bump

Thanks a lot! It worked after I generated a new SSH key for it!

Cheers,
Jinjin

**Comment by Gene233** [comment]
Hi @HelenaLC , I got an error message from build saying that the maintainer must subscribe to the bioc-devel mailing list, I submitted it but did not receive a confirmation email. How can I fix this? And once my subscription is confirmed, how can I rebuild a report without adding a bump patch (since I didn't change anything, just updated the subscription)?

Sorry to be bothering,
Jinjin

**Comment by lshep** [comment]
I have manually added you to the bioc-devel mailing list.

**Comment by HelenaLC** [comment]
Thanks Lori. @Gene233 regarding your second question, you do need to push a version bump regardless in order to trigger a rebuild. 
If you want to avoid these type of ‘wasted’ bumps in the future, it’s best to run checks locally before pushing to Bioc. (Though note that here, it doesn’t hurt as the pkg will be set to 1.0.0 with the 1st release... but still a habit worth getting accustomed to to avoid breaking changes in the future...)

**Comment by Gene233** [comment]
> Thanks Lori. @Gene233 to your second question: you need to do a version jump in any case to trigger a rebuild.

Got it! And thanks @lshep  for adding me!

**Comment by Gene233** [comment]
Hello @HelenaLC , I found a warning from BiocCheck under R4.3 that did not occur under R4.2: 
```
* Checking man page documentation...
    * WARNING: Empty or missing \value sections found in man pages.
```
I checked my man files manually, all functions have a valid return value except my dataset description man files. Does this mean that in R4.3 I also have to give a return value to each in-built dataset?

**Comment by Gene233** [comment]
> Thanks Lori. @Gene233 regarding your second question, you do need to push a version bump regardless in order to trigger a rebuild. If you want to avoid these type of ‘wasted’ bumps in the future, it’s best to run checks locally before pushing to Bioc. (Though note that here, it doesn’t hurt as the pkg will be set to 1.0.0 with the 1st release... but still a habit worth getting accustomed to to avoid breaking changes in the future...)

Thank you very much for your advice!

**Comment by HelenaLC** [comment]
First round of comments below! The first and second being the most crucial ones... Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!

# general

- [ ] Bioconductor places a strong emphasis on re-use of code and classes where it makes sense, with the general recommendation is to avoid creating methods for classes that are not your own. In this context, defining S4Methods for `SeuratObject`s doesn't seem advantageous as `Seurat` is not in Bioc, adds a *lot* of additional dependencies (it took me quite a while to actually install it...), and frequently introduced breaking changes, making your package high maintenance and susceptible to breaking at any point in time.  
You already implemented methods for plain `matrix` and various Bioc classes including `SummarizedExperiment`. If users wish to do so, this can be converted to a `SeuratObject` with `as.Seurat()`; alternatively, methods can always be run on a matrix extracted from the `SeuratObject`. You could demonstrate this in the vignette (i.e., moving `Seurat` to `Suggests:`).  
Somewhat related, if you do not need it's analysis functionality, perhaps consider whether it would be sufficient to have the `SeuratObject` package under `Suggests`, which I believe defines the class, is lighter in terms of dependencies, and presumably more stable.

- [ ] *pseudo_samples.R* caught my attention... there is a lot of existing infrastructure for these types of analysis that are worth reusing. E.g., `scuttle` has a one-line function `AggregateAcrossCells` that very efficiently does, I believe, what is implemented here. Please consider making use of this and possible other functions. I think having a look at `scuttle` might go a long way, as it was designed to contain utilities to aid others in package development.  
To give another example, in *plot.R*, `pca_matrix_plot_init` contains a lot of code to perform and visualize PCA. However, all of this is already contained in `scuttle` (and, of course, many other places). I generally worry about re-implementing to many things "from scratch". This is making your life (as a developer) hard, and is also not necessarily in the interest of your users. Perhaps consider simplifying the package structure by generally leveraging more of what's out there already, well implemented, and stable over time. 

# code

- [ ] Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).

- [ ] Generics should go in a separate script titled *AllGenerics.R* (see [here](https://contributions.bioconductor.org/r-code.html?q=allgenerics#rcode-file-names)).

- [ ] `Seurat` is not part of Bioconductor, has a large number of dependencies, and frequently introduced breaking changes. For these (and other reasons), I'd strongly recommend against depending on it. Users can easily convert a `SingleCellExperiment` to a `SeuratObject` via `Seurat::as.SingleCellExperiment()` if they wish to.

- [ ] Avoid the use of `paste` in condition signals; `message(paste(a, b, c))` can be written as just `message(a, b, c)`.

- [ ] Unit test coverage is at 80%, however, there are only 1-20 hits/line, meaning that every line of code is only minimally checked, sometimes just once... Especially with a package of this size, it is most advisable to achieve higher coverage overall - you'll be grateful in the long run.  
Specifically, I think implementing methods for so many signatures is making this especially hard. Perhaps simplifying according to 1) comment re `Seurat` and 2) comment re `scuttle` will make it easier / less work to "dive deeper" into unit testing the package more in-depth.

# vignette

- [ ] We encourage using `BiocStyle::html_document` as rendering target.

- [ ] `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.

- [ ] You are usually using code-style to highlight R code (variable, function and package names), but not always (e.g., sometimes mastR is just plain text, and objects are enclosed in normal ticks ('...') instead of back-ticks). Would be nice to be consistent here! 

- [ ] `eval=FALSE` flags and similar undermine the benefit of vignettes and are generally not allowed (see [here](http://contributions.bioconductor.org/docs.html?q=eval=FALSE#vignettes) for details).

# other

- [ ] Please update R version dependency from 4.2 to 4.3.0.

- [ ] Please keep the NEWS file on track. It currently state v0.9.0, which never existed.

- [ ] The scripts in `data-raw/` should go in `extdata/inst/scripts` (see [here](https://contributions.bioconductor.org/data.html#raw-data-and-the-instextdata-directory)).

- [ ] The `.Rhistory` should probably be removed / not be checked into GitHub and Bioc in the first place... I see that it is listed under `.gitignore`, but it still pops up for me.

**Comment by Gene233** [comment]
> First round of comments below! The first and second being the most crucial ones... Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!
> 
> # general
> * [ ]  Bioconductor places a strong emphasis on re-use of code and classes where it makes sense, with the general recommendation is to avoid creating methods for classes that are not your own. In this context, defining S4Methods for `SeuratObject`s doesn't seem advantageous as `Seurat` is not in Bioc, adds a _lot_ of additional dependencies (it took me quite a while to actually install it...), and frequently introduced breaking changes, making your package high maintenance and susceptible to breaking at any point in time.
>   You already implemented methods for plain `matrix` and various Bioc classes including `SummarizedExperiment`. If users wish to do so, this can be converted to a `SeuratObject` with `as.Seurat()`; alternatively, methods can always be run on a matrix extracted from the `SeuratObject`. You could demonstrate this in the vignette (i.e., moving `Seurat` to `Suggests:`).
>   Somewhat related, if you do not need it's analysis functionality, perhaps consider whether it would be sufficient to have the `SeuratObject` package under `Suggests`, which I believe defines the class, is lighter in terms of dependencies, and presumably more stable.
> * [ ]  _pseudo_samples.R_ caught my attention... there is a lot of existing infrastructure for these types of analysis that are worth reusing. E.g., `scuttle` has a one-line function `AggregateAcrossCells` that very efficiently does, I believe, what is implemented here. Please consider making use of this and possible other functions. I think having a look at `scuttle` might go a long way, as it was designed to contain utilities to aid others in package development.
>   To give another example, in _plot.R_, `pca_matrix_plot_init` contains a lot of code to perform and visualize PCA. However, all of this is already contained in `scuttle` (and, of course, many other places). I generally worry about re-implementing to many things "from scratch". This is making your life (as a developer) hard, and is also not necessarily in the interest of your users. Perhaps consider simplifying the package structure by generally leveraging more of what's out there already, well implemented, and stable over time.
> 
> # code
> * [ ]  Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).
> * [ ]  Generics should go in a separate script titled _AllGenerics.R_ (see [here](https://contributions.bioconductor.org/r-code.html?q=allgenerics#rcode-file-names)).
> * [ ]  `Seurat` is not part of Bioconductor, has a large number of dependencies, and frequently introduced breaking changes. For these (and other reasons), I'd strongly recommend against depending on it. Users can easily convert a `SingleCellExperiment` to a `SeuratObject` via `Seurat::as.SingleCellExperiment()` if they wish to.
> * [ ]  Avoid the use of `paste` in condition signals; `message(paste(a, b, c))` can be written as just `message(a, b, c)`.
> * [ ]  Unit test coverage is at 80%, however, there are only 1-20 hits/line, meaning that every line of code is only minimally checked, sometimes just once... Especially with a package of this size, it is most advisable to achieve higher coverage overall - you'll be grateful in the long run.
>   Specifically, I think implementing methods for so many signatures is making this especially hard. Perhaps simplifying according to 1) comment re `Seurat` and 2) comment re `scuttle` will make it easier / less work to "dive deeper" into unit testing the package more in-depth.
> 
> # vignette
> * [ ]  We encourage using `BiocStyle::html_document` as rendering target.
> * [ ]  `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.
> * [ ]  You are usually using code-style to highlight R code (variable, function and package names), but not always (e.g., sometimes mastR is just plain text, and objects are enclosed in normal ticks ('...') instead of back-ticks). Would be nice to be consistent here!
> * [ ]  `eval=FALSE` flags and similar undermine the benefit of vignettes and are generally not allowed (see [here](http://contributions.bioconductor.org/docs.html?q=eval=FALSE#vignettes) for details).
> 
> # other
> * [ ]  Please update R version dependency from 4.2 to 4.3.0.
> * [ ]  Please keep the NEWS file on track. It currently state v0.9.0, which never existed.
> * [ ]  The scripts in `data-raw/` should go in `extdata/inst/scripts` (see [here](https://contributions.bioconductor.org/data.html#raw-data-and-the-instextdata-directory)).
> * [ ]  The `.Rhistory` should probably be removed / not be checked into GitHub and Bioc in the first place... I see that it is listed under `.gitignore`, but it still pops up for me.

Thank you very much for your quick comments! I'll have a quick check and feedback ASAP! I'll try to get higher coverage for unit tests and consider on `Seurat ` class. Once I get progress, I'll comment here and update you! 
Thank you very much!

Cheers,
Jinjin

**Comment by Gene233** [comment]
> First round of comments below! The first and second being the most crucial ones... Note that many are suggestions and not strictly required. Please comment back here with what has/not been addressed how/why not, or any comments regarding any needed clarification or questions you might have; happy to help/discuss. Cheers!
> 
> # general
> * [x]  Bioconductor places a strong emphasis on re-use of code and classes where it makes sense, with the general recommendation is to avoid creating methods for classes that are not your own. In this context, defining S4Methods for `SeuratObject`s doesn't seem advantageous as `Seurat` is not in Bioc, adds a _lot_ of additional dependencies (it took me quite a while to actually install it...), and frequently introduced breaking changes, making your package high maintenance and susceptible to breaking at any point in time.
>   You already implemented methods for plain `matrix` and various Bioc classes including `SummarizedExperiment`. If users wish to do so, this can be converted to a `SeuratObject` with `as.Seurat()`; alternatively, methods can always be run on a matrix extracted from the `SeuratObject`. You could demonstrate this in the vignette (i.e., moving `Seurat` to `Suggests:`).
>   Somewhat related, if you do not need it's analysis functionality, perhaps consider whether it would be sufficient to have the `SeuratObject` package under `Suggests`, which I believe defines the class, is lighter in terms of dependencies, and presumably more stable.
> * [ ]  _pseudo_samples.R_ caught my attention... there is a lot of existing infrastructure for these types of analysis that are worth reusing. E.g., `scuttle` has a one-line function `AggregateAcrossCells` that very efficiently does, I believe, what is implemented here. Please consider making use of this and possible other functions. I think having a look at `scuttle` might go a long way, as it was designed to contain utilities to aid others in package development.
>   To give another example, in _plot.R_, `pca_matrix_plot_init` contains a lot of code to perform and visualize PCA. However, all of this is already contained in `scuttle` (and, of course, many other places). I generally worry about re-implementing to many things "from scratch". This is making your life (as a developer) hard, and is also not necessarily in the interest of your users. Perhaps consider simplifying the package structure by generally leveraging more of what's out there already, well implemented, and stable over time.
> 
> # code
> * [x]  Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).
> * [ ]  Generics should go in a separate script titled _AllGenerics.R_ (see [here](https://contributions.bioconductor.org/r-code.html?q=allgenerics#rcode-file-names)).
> * [ ]  `Seurat` is not part of Bioconductor, has a large number of dependencies, and frequently introduced breaking changes. For these (and other reasons), I'd strongly recommend against depending on it. Users can easily convert a `SingleCellExperiment` to a `SeuratObject` via `Seurat::as.SingleCellExperiment()` if they wish to.
> * [x]  Avoid the use of `paste` in condition signals; `message(paste(a, b, c))` can be written as just `message(a, b, c)`.
> * [ ]  Unit test coverage is at 80%, however, there are only 1-20 hits/line, meaning that every line of code is only minimally checked, sometimes just once... Especially with a package of this size, it is most advisable to achieve higher coverage overall - you'll be grateful in the long run.
>   Specifically, I think implementing methods for so many signatures is making this especially hard. Perhaps simplifying according to 1) comment re `Seurat` and 2) comment re `scuttle` will make it easier / less work to "dive deeper" into unit testing the package more in-depth.
> 
> # vignette
> * [ ]  We encourage using `BiocStyle::html_document` as rendering target.
> * [ ]  `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.
> * [x]  You are usually using code-style to highlight R code (variable, function and package names), but not always (e.g., sometimes mastR is just plain text, and objects are enclosed in normal ticks ('...') instead of back-ticks). Would be nice to be consistent here!
> * [ ]  `eval=FALSE` flags and similar undermine the benefit of vignettes and are generally not allowed (see [here](http://contributions.bioconductor.org/docs.html?q=eval=FALSE#vignettes) for details).
> 
> # other
> * [x]  Please update R version dependency from 4.2 to 4.3.0.
> * [x]  Please keep the NEWS file on track. It currently state v0.9.0, which never existed.
> * [x]  The scripts in `data-raw/` should go in `extdata/inst/scripts` (see [here](https://contributions.bioconductor.org/data.html#raw-data-and-the-instextdata-directory)).
> * [x]  The `.Rhistory` should probably be removed / not be checked into GitHub and Bioc in the first place... I see that it is listed under `.gitignore`, but it still pops up for me.

Hi @HelenaLC , thank you very much for your feedback. 
# general
As for the first point of General, I aggree it might be enough to use `SeuratObject` instead of `Seurat`. As for `Suggests`, I tried it before, and it proved if I want to use `Seurat` as accepted input format, I have to import it instead of suggest it.

However, for the codes re-usage, I also checked aggregating functions and pca plot functions, but none of them fully implemented what I want to do. The `pseudo_sample` function in mastR can not only aggregate cells based on factors, but also divide large pseudo-samples into smaller pseudo samples of comparable size by using the `min.cell` and `max.cell` parameters. The cell size of pseudo sample will be a random number within the range from `min.cells` to `max.cells`. This allows to generate pseudo samples with comparable lib.sizes, all replicates from one factorial combination will have a number suffix, which can then be used to obtain the co-linearity across replicates. While aggregation function in scuttle directly aggregates all cells of one combination into one pseudo sample, which always resulting highly variable lib.sizes.

And another reason is that I don't want to import too many packages if I just want to use 1 or 2 functions of them. So I implemented my own version of these functions.

# code
For codes, I'll move all generic functions under 'AllGeneric.R', and leave class definitions under 'AllClass.R', following the guideline.

For `Seurat`, the same question, do you think it's OK to leave `SeuratObject` as a dependent package?

I will check all conditions and avoid using paste in it.

For the coverage, I'm a bit confused, I actually tried to make a unit test for each single function, for the generic function, I even tested with different input classes. Would you mind giving me more details on how to further improve the coverage? I'm sorry if it's a naive question.

# vignette
I tried `BiocStyle` before, but then I got a warning that the package size is over 5Mb due to the large size of vignette. So I gave up using `BiocStyle` in vignette. I would appreciate if you could give me some tricks to reduce its size.

For `eval=FALSE` flags, I used it mainly for 1) PanglaoDB related functions, as it requires internet connection to the latest website; and 2) applications of final signatures,  I didn't evaluate them, as they are not part of the mastR workflow, but just demonstrating how to use the results of mastR. Do you think I should keep the application part, but set `eval=TRUE` or just delete it?

# other
I have fixed everything in this section. But since our server is under maintenance, I can only push all my changes once the server is available again after April 8.

# summary
These are some of my initial discussions in response to your feedback, and I'm glad to discuss specific details of the changes with you.

Hope you have a great Easter holiday!

Cheers,
Jinjin

**Comment by Gene233** [comment]
Hi @HelenaLC , I have pushed the latest version on Bioconductor.
I hope it looks better and I am happy to discuss more details with you (which I mentioned in my last comment).

Cheers,
Jinjin

**Comment by HelenaLC** [comment]
- Yes, keeping `SeuratObject` under `Suggests:` is a lot better! It is a lot lighter (in terms of dependencies) than `Seurat`, and all you need here to define S4Methods for the class.

- Re unit tests: Apologies if I wasn't clear enough. What I meant is that yes, you do have fairly high coverage (>80%), and all functions are being tested. But besides coverage, there is unit test **depth**, i.e., how often bits of code are being tested. This is currently at 1 hit/line for some scripts... you can see how often every line of code is being tests via, e.g., `devtools::test_coverage()`:
![image](https://user-images.githubusercontent.com/14542264/231111978-20e6c07d-009d-4921-adb8-cba893597eae.png)

- [ ] Re `BiocStyle`: The vignette doesn't seem large at all compare to other packages (e.g., [here](https://www.bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/differential.html) is a vignette from a package of mine, there is even a 2nd vignette, and still size is not an issue). It's currently less than 1MB for me and, to the best of my knowledge, changing style shouldn't increase that significantly. Perhaps you could try again and, if need be, reduce the size/resolution of some large figures? Though I don't think that should be necessary... 

- [ ] When the above works, just a remember that: `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.

- [ ] As mentioned previously: Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).

- [ ] There is still a lot of inconsistency in the vignette re code-style... Please have a quick read and use code-style for any package, function, variables names - this greatly help readability, and to distinguish between method/technology/... names vs. software and other R-related references. 

- [ ] Re `eval=FALSE` in the vignette: I don't see any harm in running parts oft the analysis that are not specific to your package. Many other workflows do this, e.g., to demonstrate upstream preprocessing or downstream analyses steps. So yes, please remove these flags or otherwise (if need be) remove the code chunks and use descriptive text/pointers to replace them.

- [ ] Please take a few minutes to keep the NEWS file on track. It now starts with v0.99.6 and does not at all cover all the changes that have been made. Remember this help users understand

**Comment by Gene233** [comment]
> * Yes, keeping `SeuratObject` under `Suggests:` is a lot better! It is a lot lighter (in terms of dependencies) than `Seurat`, and all you need here to define S4Methods for the class.
> * Re unit tests: Apologies if I wasn't clear enough. What I meant is that yes, you do have fairly high coverage (>80%), and all functions are being tested. But besides coverage, there is unit test **depth**, i.e., how often bits of code are being tested. This is currently at 1 hit/line for some scripts... you can see how often every line of code is being tests via, e.g., `devtools::test_coverage()`:
>   ![image](https://user-images.githubusercontent.com/14542264/231111978-20e6c07d-009d-4921-adb8-cba893597eae.png)
> * [x]  Re `BiocStyle`: The vignette doesn't seem large at all compare to other packages (e.g., [here](https://www.bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/differential.html) is a vignette from a package of mine, there is even a 2nd vignette, and still size is not an issue). It's currently less than 1MB for me and, to the best of my knowledge, changing style shouldn't increase that significantly. Perhaps you could try again and, if need be, reduce the size/resolution of some large figures? Though I don't think that should be necessary...
> * [x]  When the above works, just a remember that: `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.
> * [x]  As mentioned previously: Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).
> * [x]  There is still a lot of inconsistency in the vignette re code-style... Please have a quick read and use code-style for any package, function, variables names - this greatly help readability, and to distinguish between method/technology/... names vs. software and other R-related references.
> * [x]  Re `eval=FALSE` in the vignette: I don't see any harm in running parts oft the analysis that are not specific to your package. Many other workflows do this, e.g., to demonstrate upstream preprocessing or downstream analyses steps. So yes, please remove these flags or otherwise (if need be) remove the code chunks and use descriptive text/pointers to replace them.
> * [x]  Please take a few minutes to keep the NEWS file on track. It now starts with v0.99.6 and does not at all cover all the changes that have been made. Remember this help users understand

Hi @HelenaLC , Thank you very much for your patient explaination! 
* Test Coverage
I appreciate your feedback on the need for more comprehensive testing in future maintenance efforts. Due to the limited nature of some functions being solely for visualization purposes, deep unit tests were not initially included. However, considering the upcoming deadline, could I please ask for your understanding and permission to work on the deeper tests later, after the release is done?

* Vignette
I have reformatted the vignette using `BiocStyle`. I have also fixed the inconsistency issue by enclosing all functions, packages, parameters and variables in backticks (``) and all class names and other special terms in single quotes (').
Thank you for your comment on the `eval=FALSE` issue. I understand your concern, but I used this option because the PanglaoDB functions require an Internet connection to work. I wanted to avoid building the vignette with this dependency, as it might cause some instability or errors. Regarding the applications section, I am aware that the annotation application may take a bit longer time to build the vignette which results in the time for building vignette to be > 2 mins. And there're some packages for application but non-relevent to my package will be added into `Suggests` and increase the time of installing dependencies. Is that still acceptable?

If you believe it would be appropriate, I would be happy to push it to bioc.

Cheers,
Jinjin

**Comment by Gene233** [comment]
Hi @HelenaLC , I pushed my changes to bioc and set all the chunks to run except the PanglaoDB section (because of the internet connection concerns I mentioned last time). Do you think it is acceptable to add the packages needed for the applications in `Suggests`? Or shall I just remove the whole application section and maybe write it to another separate file if needed?

Thank you very much!
Jinjin

**Comment by HelenaLC** [comment]
- Yes, keeping `SeuratObject` under `Suggests:` is a lot better! It is a lot lighter (in terms of dependencies) than `Seurat`, and all you need here to define S4Methods for the class.

- Re unit tests: Yes, these are not a requirement, though we very much like to see them. And I encourage you to keep working on them to build a solid test base for the future! But it's not a necessity for acceptance at this point.

- [ ] Re `BiocStyle`: The vignette doesn't seem large at all compare to other packages (e.g., [here](https://www.bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/differential.html) is a vignette from a package of mine, there is even a 2nd vignette, and still size is not an issue). It's currently less than 1MB for me and, to the best of my knowledge, changing style shouldn't increase that significantly. Perhaps you could try again and, if need be, reduce the size/resolution of some large figures? Though I don't think that should be necessary... 

- [ ] When the above works, just a remember that: `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.

- [ ] As mentioned previously: Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).

- [ ] There is still a lot of inconsistency in the vignette re code-style... Please have a quick read and use code-style for any package, function, variables names - this greatly help readability, and to distinguish between method/technology/... names vs. software and other R-related references. 

- [ ] Re `eval=FALSE` in the vignette: I don't see any harm in running parts oft the analysis that are not specific to your package. Many other workflows do this, e.g., to demonstrate upstream preprocessing or downstream analyses steps. So yes, please remove these flags or otherwise (if need be) remove the code chunks and use descriptive text/pointers to replace them.

- [ ] Please take a few minutes to keep the NEWS file on track. It now starts with v0.99.6 and does not at all cover all the changes that have been made. Remember this help users understand

**Comment by Gene233** [comment]
> * Yes, keeping `SeuratObject` under `Suggests:` is a lot better! It is a lot lighter (in terms of dependencies) than `Seurat`, and all you need here to define S4Methods for the class.
> * Re unit tests: Yes, these are not a requirement, though we very much like to see them. And I encourage you to keep working on them to build a solid test base for the future! But it's not a necessity for acceptance at this point.
> * [x]  Re `BiocStyle`: The vignette doesn't seem large at all compare to other packages (e.g., [here](https://www.bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/differential.html) is a vignette from a package of mine, there is even a 2nd vignette, and still size is not an issue). It's currently less than 1MB for me and, to the best of my knowledge, changing style shouldn't increase that significantly. Perhaps you could try again and, if need be, reduce the size/resolution of some large figures? Though I don't think that should be necessary...
> * [x]  When the above works, just a remember that: `BiocStyle` has designated functions to highlight and hyperlink external package, e.g., `r Biocpkg("...")` (there are similar commands for CRAN and GitHub packages). This is nice in that it directly directs users to external packages.
> * [x]  As mentioned previously: Functions that call functions from packages under `Suggests:` should check whether or not these are available at the very first (see [here](https://contributions.bioconductor.org/description.html?q=suggests#depends-imports-suggests-enhances) for details).
> * [x]  There is still a lot of inconsistency in the vignette re code-style... Please have a quick read and use code-style for any package, function, variables names - this greatly help readability, and to distinguish between method/technology/... names vs. software and other R-related references.
> * [x]  Re `eval=FALSE` in the vignette: I don't see any harm in running parts oft the analysis that are not specific to your package. Many other workflows do this, e.g., to demonstrate upstream preprocessing or downstream analyses steps. So yes, please remove these flags or otherwise (if need be) remove the code chunks and use descriptive text/pointers to replace them.
> * [x]  Please take a few minutes to keep the NEWS file on track. It now starts with v0.99.6 and does not at all cover all the changes that have been made. Remember this help users understand

Hi @HelenaLC , thank you for your updates. I've done almost all of above. 
* I have used backticks (``) to highlight all functions, packages, parameters and variables, and single quotes (') to highlight all class names and other special terms.
* I have set all code chunks to eval=TRUE, except for the one that uses PanglaoDB, because it requires an Internet connection that may cause instability and error.
* I have tried to suggest the SeuratObject package instead of importing it, but I encountered a warning that the "Seurat" class was not found. The warning disappeared when I imported the SeuratObject package.
* I have updated the NEWS file to document the changes made for each bump.

I hope these changes are satisfactory. Please let me know if you have any feedback or suggestions. Thank you for your time and attention.

Cheers,
Jinjin

**Comment by HelenaLC** [comment]
Thanks for the quick fix and addressing reviews. I'd appreciate if you could take a couple minutes to do the following ... but looking pretty good to me now!
- As mentioned previously, it would be nice if you could link to external GitHub/CRAN/Bioc packages using the designated `BiocStyle` functions. E.g., line 881 "`splatter` package" -> `BiocStyle::Biocpkg("splatter")`. For CRAN/GitHub packages, you'd use `BiocStyle::CRANpkg()` and `BiocStyle::GitHubpkg`, respectively.
- Still some inconsistencies here and there re code-style... e.g., line 288+ "'GeneSet' object" should be `GeneSet` since it's an R class; also line 190, 337... perhaps doing a search for single ticks ' could help fixing these.
- There's still a NOTE in the build report re using `paste` in condition signals. Perhaps doing a global search for `message(` and `warning(` would help to remove such occurrences.

**Comment by Gene233** [comment]
> Thanks for the quick fix and addressing reviews. I'd appreciate if you could take a couple minutes to do the following ... but looking pretty good to me now!
> 
> * As mentioned previously, it would be nice if you could link to external GitHub/CRAN/Bioc packages using the designated `BiocStyle` functions. E.g., line 881 "`splatter` package" -> `BiocStyle::Biocpkg("splatter")`. For CRAN/GitHub packages, you'd use `BiocStyle::CRANpkg()` and `BiocStyle::GitHubpkg`, respectively.
> * Still some inconsistencies here and there re code-style... e.g., line 288+ "'GeneSet' object" should be `GeneSet` since it's an R class; also line 190, 337... perhaps doing a search for single ticks ' could help fixing these.
> * There's still a NOTE in the build report re using `paste` in condition signals. Perhaps doing a global search for `message(` and `warning(` would help to remove such occurrences.

Hi @HelenaLC , sorry to be late to reply.

I appreciate your patience and timely feedback. I have made the changes you suggested, such as using BiocStyle functions for external links and formatting the R class with ``. I also resolved the note about paste in conditions. Please let me know if everything looks fine now.

Thank you very much for your guidance!

Jinjin

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/Gene233.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("mastR")`. The package 'landing page' will be created at

https://bioconductor.org/packages/mastR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
